apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ServletApiAdvice.java

```diff
@@ -100,49 +122,96 @@ public static void onEnterServletService(@Advice.Argument(0) ServletRequest serv
+                        ServletInstrumentation.servletTransactionCreationHelperManager.getForClassLoaderOfClass(HttpServletRequest.class);
+                    if (helper != null) {
+                        transaction = helper.createAndActivateTransaction(request);
+                    }
                 }
-            }
 
-            if (transaction == null) {
-                // if the request is excluded, avoid matching all exclude patterns again on each filter invocation
-                excluded.set(Boolean.TRUE);
-                return;
-            }
-            final Request req = transaction.getContext().getRequest();
-            if (transaction.isSampled() && tracer.getConfig(CoreConfiguration.class).isCaptureHeaders()) {
-                if (request.getCookies() != null) {
-                    for (Cookie cookie : request.getCookies()) {
-                        req.addCookie(cookie.getName(), cookie.getValue());
+                if (transaction == null) {
+                    // if the request is excluded, avoid matching all exclude patterns again on each filter invocation
+                    excluded.set(Boolean.TRUE);
+                    return;
+                }
+                final Request req = transaction.getContext().getRequest();
+                if (transaction.isSampled() && tracer.getConfig(CoreConfiguration.class).isCaptureHeaders()) {
+                    if (request.getCookies() != null) {
+                        for (Cookie cookie : request.getCookies()) {
+                            req.addCookie(cookie.getName(), cookie.getValue());
+                        }
+                    }
+                    final Enumeration<String> headerNames = request.getHeaderNames();
+                    if (headerNames != null) {
+                        while (headerNames.hasMoreElements()) {
+                            final String headerName = headerNames.nextElement();
+                            req.addHeader(headerName, request.getHeaders(headerName));
+                        }
                     }
                 }
-                final Enumeration<String> headerNames = request.getHeaderNames();
-                if (headerNames != null) {
-                    while (headerNames.hasMoreElements()) {
-                        final String headerName = headerNames.nextElement();
-                        req.addHeader(headerName, request.getHeaders(headerName));
+                transaction.setFrameworkName(FRAMEWORK_NAME);
+
+                servletTransactionHelper.fillRequestContext(transaction, request.getProtocol(), request.getMethod(), request.isSecure(),
+                    request.getScheme(), request.getServerName(), request.getServerPort(), request.getRequestURI(), request.getQueryString(),
+                    request.getRemoteAddr(), request.getHeader("Content-Type"));
+            } else if (transaction == null) {
```

Comment:
Why is the `transaction == null` required here? Seems it will always be `null`

Reply:
i changed here condition to -> 
```
servletRequest.getDispatcherType() != DispatcherType.ASYNC
```